### DISCUSSION TITLE
AttributeError: module 'SCHAModules' has no attribute 'thermodynamic'

### DISCUSSION THREAD

**Discussion by Tongca** [SSCHAcode/python-sscha#184] [Q&A] [answered]
Dear developers,

When I try to start a sscha calculation or import SCHAModules I get AttributeError: module 'SCHAModules' has no attribute 'thermodynamic' The compilation seems to end well, everything is installed and I can load both cellconstructor and sscha. However, the problem is in SCHAModules as it is not seeing the thermodynamic attribute: Do you have any suggestion on how to solve it?

Thanks a lot!
Yue

**Comment by mesonepigreco** [comment]
Hi, yes we just received another user who reported a similar problem (issue #183 )
Can you give some details on how you installed the code? Was it on your local pc or on a cluster? Did you use anaconda or the builtin python version? 
Thanks for your report

**Reply by Tongca** [reply]
Thank you for your reply. We installed sscha using anaconda from a cluster. 
conda create -n sscha -c conda-forge python=3.10 gfortran libblas lapack openmpi julia openmpi-mpicc pip numpy scipy spglib
conda activate sscha
pip install ase julia mpi4py
pip install cellconstructor python-sscha tdscha
All compilations look great. However, I am unable to import the thermodynamic module.

Python 3.10.13 | packaged by conda-forge | (main, Dec 23 2023, 15:36:39) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sscha
>>> import SCHAModules
>>> import thermodynamic
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'thermodynamic'

**Comment by diegomartinez2** [comment]
I'll point out that "module thermodynamic" is in 'module_new_termodynamic.f90', so if the fortran compilation failed, there will be no module to import.
But, I do not think that's the problem because Tongca used the pip install and that sould gave an error message if the fortran compilation failed.

**Reply by Tongca** [reply]
I agree with your point of view, I tried to modify module_new_thermodynamic" to 'module_termodynamic.f90, and modified setup.py and the same problem still occurred.

**Reply by diegomartinez2** [reply]
So you used the manual installation with 'python setup.py install --record files.txt 2>&1 | tee install.log'?

---manual compilation of python-sscha (only for the local user)---
cd python-sscha
python3 setup.py install --user
---

**Reply by Tongca** [reply]
Yes, I have tried both using anaconda and manual installation, I will provide an install.log
[install.log](https://github.com/SSCHAcode/python-sscha/files/14006528/install.log)

**Reply by diegomartinez2** [reply]
There is this message in install.log: 
"Constructing F90 module support for "thermodynamic"...
			Skipping thermodynamic since it is in 'use'..."

Does that implies that the wrapper function is not created? 
(Is there a 'thermodynamic' named python module already there? No, because "No module named 'thermodynamic'" error message at import. I'm confused. That's the same message [DromaninCM](https://github.com/DromaninCM) got -- issue https://github.com/SSCHAcode/python-sscha/issues/183 --.)

**Reply by Tongca** [reply]
I tried to import SCHAModules and it seemed that all modules were compiled.

Python 3.10.13 | packaged by conda-forge | (main, Dec 23 2023, 15:36:39) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import SCHAModules
>>> print(dir(SCHAModules))
['_SCHAModules_error', '__doc__', '__f2py_numpy_version__', '__file__', '__loader__', '__name__', '__package__', '__spec__', '__version__', 'anharmonic', 'cell_force', 'get_cmat', 'get_emat', 'get_fmunu', 'get_g', 'get_gradient_supercell', 'get_gradient_supercell_fast', 'get_gradient_supercell_new', 'get_odd_straight', 'get_odd_straight_with_v4', 'get_stress_tensor', 'get_unpreconditioned_gradient', 'get_upsilon_matrix', 'get_v3', 'get_v4', 'get_ymu', 'multiply_lambda_tensor', 'stochastic']

**Reply by diegomartinez2** [reply]
In my case I have also the 'thermodynamic' at the end:
>>> import SCHAModules
>>> print(dir(SCHAModules)) 
['_SCHAModules_error', '__doc__', '__f2py_numpy_version__', '__file__', '__loader__', '__name__', '__package__', '__sp
ec__', '__version__', 'anharmonic', 'cell_force', 'get_cmat', 'get_emat', 'get_fmunu', 'get_g', 'get_gradient_supercel
l', 'get_gradient_supercell_fast', 'get_gradient_supercell_new', 'get_odd_straight', 'get_odd_straight_with_v4', 'get_
stress_tensor', 'get_unpreconditioned_gradient', 'get_upsilon_matrix', 'get_v3', 'get_v4', 'get_ymu', 'multiply_lambda
_tensor', 'stochastic', 'thermodynamic']

**Reply by Tongca** [reply]
I reinstalled sscha using anaconda in another cluster and the thermodynamics module appeared, but the problem still occurs in the current cluster.

**Reply by mesonepigreco** [reply]
It is really weird. Maybe the module is referenced incorrectly in other submodules, and this causes it to fail to be included in the shared library in some environments. Module thermodynamics is not so big and is used little in the code. Maybe we can avoid calling it in the Python code, at least the main version.

@Tongca can you report exactly the error when you try to execute the sscha? 

In this way, we can help you circumvent the problem and run the calculation without using that module.

**Reply by diegomartinez2** [reply]
Hi Lorenzo, a pythonized version of that fortran code will be something like:
----------------- Thermodynamic.py ------------------------------------------
```
import numpy as np
from scipy.special import tanh, sinh, cosh

def w_to_a(w, T, n):
    a = np.zeros(n)
    if T == 0.0:
        a[:] = np.sqrt(1.0 / (2.0 * w))
    else:
        a[:] = np.sqrt((1.0 / tanh(0.5 * w * 315774.65221921849 / T)) / (2.0 * w))
    return a

def w_to_da(w, T, n):
    da = np.zeros(n)
    if T == 0.0:
        da[:] = -np.sqrt(1.0 / (8.0 * w**3.0))
    else:
        beta = 315774.65221921849 / T
        a = w * beta + sinh(w * beta)
        b = np.sqrt(1.0 / (32.0 * w**3.0 * (sinh(0.5 * w * beta)**3.0) * cosh(0.5 * w * beta)))
        da[:] = -a * b
    return da

def dW_f0_u0(w, T):
    if T == 0.0:
        return 0.25
    else:
        return 0.25 * (2.0 * nb(w, T) + 1.0 + 2.0 * (315774.65221921849 / T) * w * np.exp(w * 315774.65221921849 / T) * nb(w, T)**2.0)

def nb(w, T):
    if T == 0.0:
        return 0.0
    else:
        return 1.0 / (np.exp(w * 315774.65221921849 / T ) - 1.0)
```
-----------------------------------
If we put that into the python-sscha/Modules and edit the setup to take out the fortran version, it may work as expected.

EDIT: It need a little more to work... change the SSCHAmodules calls in "Ensemble.py", etc.
EDIT: sed -i 's/"SCHAModules.thermodynamic"/"Thermodynamic"/' Ensemble.py

**Reply by diegomartinez2** [reply, marked as answer]
The patch should be (not tested yet):

1. Create the file "Thermodynamic.py" with the code above into python-sscha/Modules
```python
import numpy as np
from scipy.special import tanh, sinh, cosh

def w_to_a(w, T):
    n = len(w)
    a = np.zeros(n)
    if T == 0.0:
        a[:] = np.sqrt(1.0 / (2.0 * w))
    else:
        a[:] = np.sqrt((1.0 / tanh(0.5 * w * 315774.65221921849 / T)) / (2.0 * w))
    return a
```
2. In python-sscha/Modules run this command: `sed -i 's/"SCHAModules.thermodynamic"/"Thermodynamic"/' Ensemble.py`
3. Add the line ``import sscha.Thermodynamic as Thermodynamic`` at the beginning of the file Ensemble.py 
~~3. Optional: Edit python-sscha/setup.py to comment the old fortran code "SCHAModules/module_new_thermodynamic.f90".~~
4. run python setup.py install --record files.txt 2>&1 | tee install.log

**Reply by mesonepigreco** [reply]
This seems a solution. I link here also issue #183

**Reply by mesonepigreco** [reply]
@diegomartinez2 It's probably even easier, as the module thermodynamics only exports the w_to _a and w_to_da subroutines. Moreover, the only subroutine that is called from Ensemble is w_to_a, so we just need that very simple function. Moreover, we do not need the last n argument, as that is the length of w, (the python bindings automatically takes care of it).

That is, indeed, if the rest of the SCHAModules has been compiled and linked correctly against those subroutines and the only problem is in the creation of the python bindings of the module (which seems the case)

**Comment by DromaninCM** [comment]
> The patch should be (not tested yet):
> 
> 1. Create the file "Thermodynamic.py" with the code above into python-sscha/Modules
> 2. In python-sscha/Modules run this command: `sed -i 's/"SCHAModules.thermodynamic"/"Thermodynamic"/' Ensemble.py`
> 3. Optional: Edit python-sscha/setup.py to comment the old fortran code "SCHAModules/module_new_thermodynamic.f90".
> 4. run python setup.py install --record files.txt 2>&1 | tee install.log

Dear all,

I tried doing this but now the python-sscha is not compiling:

INFO: gfortran:f90: SCHAModules/module_anharmonic.f90
SCHAModules/module_anharmonic.f90:3.6:

  use thermodynamic
      1
Fatal Error: Can't open module file 'thermodynamic.mod' for reading at (1): No such file or directory

I'm attaching the install.log

Davide
[install.log](https://github.com/SSCHAcode/python-sscha/files/14012735/install.log)

**Reply by mesonepigreco** [reply]
Hi Davide,
Could you remove step 3 (i.e., do not touch the original setup.py)? That is wrong. The rest of the Fortran code requires the module thermodynamic, so it should not be removed from the installation. 
Also, the new Python file Thermodynamic.py should just be:

```python
import numpy as np
from numpy import tanh, sinh, cosh

def w_to_a(w, T):
    n = len(w)
    a = np.zeros(n)
    if T == 0.0:
        a[:] = np.sqrt(1.0 / (2.0 * w))
    else:
        a[:] = np.sqrt((1.0 / tanh(0.5 * w * 315774.65221921849 / T)) / (2.0 * w))
    return a
```

It is not equivalent to the previous one (it takes one less parameter). 

Let me know if this works.

**Reply by DromaninCM** [reply]
Hi Lorenzo,

It is still not working, I did everything:

(Lorenzo_SSCHA) [romanind@ruche01 sscha_byhand]$ cd python-sscha-master/
(Lorenzo_SSCHA) [romanind@ruche01 python-sscha-master]$ cd Modules/
(Lorenzo_SSCHA) [romanind@ruche01 Modules]$ vi Thermodynamic.py
(Lorenzo_SSCHA) [romanind@ruche01 Modules]$ sed -i 's/"SCHAModules.thermodynamic"/"Thermodynamic"/' Ensemble.py
(Lorenzo_SSCHA) [romanind@ruche01 Modules]$ cd ..
(Lorenzo_SSCHA) [romanind@ruche01 python-sscha-master]$ LDFLAGS="-L/gpfs/softs/spack_0.17/opt/spack/linux-centos7-cascadelake/gcc-11.2.0/netlib-lapack-3.9.1-xu2h7uzaxe4t34uchl2msjjcd36a7ehz/lib64/" python setup.py install --record files.txt 2>&1 | tee install.log

This time the installation ends (I'm attaching the log file), but I still get:

Traceback (most recent call last):
  File "/gpfs/workdir/romanind/CNT_SSCHA/hessian.py", line 48, in <module>
    dyn_hessian = ens.get_free_energy_hessian(include_v4 = INCLUDE_V4, get_full_hessian = True,verbose = False)
  File "/gpfs/users/romanind/.conda/envs/Lorenzo_SSCHA/lib/python3.10/site-packages/sscha/Ensemble.py", line 3657, in get_free_energy_hessian
    a = SCHAModules.thermodynamic.w_to_a(w, self.current_T)
AttributeError: module 'SCHAModules' has no attribute 'thermodynamic'

Davide

[install.log](https://github.com/SSCHAcode/python-sscha/files/14022026/install.log)

**Reply by diegomartinez2** [reply]
That strange, the `sed -i 's/"SCHAModules.thermodynamic"/"Thermodynamic"/' Ensemble.py` must have changed the Ensemble.py. Did you do a `python setup.py install --record files.txt 2>&1 | tee install.log`?

In this case I will also do `sed -i 's/"SCHAModules.thermodynamic"/"Thermodynamic"/' /gpfs/users/romanind/.conda/envs/Lorenzo_SSCHA/lib/python3.10/site-packages/sscha/Ensemble.py` to change all the SCHAModules.thermodynamic calls for Thermodynamic calls. Also put the `Thermodynamic.py` there `/gpfs/users/romanind/.conda/envs/Lorenzo_SSCHA/lib/python3.10/site-packages/sscha/Thermodynamic.py`

Note that that is not the general correct way to do it, this is a quick-patch for your case.

**Reply by DromaninCM** [reply]
Dear Diego,

Even doing this doesn't work:

[romanind@ruche01 CNT_SSCHA]$ sed -i 's/"SCHAModules.thermodynamic"/"Thermodynamic"/' /gpfs/users/romanind/.conda/envs/Lorenzo_SSCHA/lib/python3.10/site-packages/sscha/Ensemble.py

[romanind@ruche01 sscha]$ cd /gpfs/users/romanind/.conda/envs/Lorenzo_SSCHA/lib/python3.10/site-packages/sscha/
[romanind@ruche01 sscha]$ ls
AdvancedCalculations.py  Cluster.py    Ensemble.py   Optimizer.py  Relax.py          Thermodynamic.py  Utilities.py  __pycache__        fourier_gradient.jl
Calculator.py            Dynamical.py  Minimizer.py  Parallel.py   SchaMinimizer.py  Tools.py          __init__.py   aiida_ensemble.py

I obtain always the error:

Traceback (most recent call last):
  File "/gpfs/workdir/romanind/CNT_SSCHA/hessian.py", line 48, in <module>
    dyn_hessian = ens.get_free_energy_hessian(include_v4 = INCLUDE_V4, get_full_hessian = True,verbose = False)
  File "/gpfs/users/romanind/.conda/envs/Lorenzo_SSCHA/lib/python3.10/site-packages/sscha/Ensemble.py", line 3657, in get_free_energy_hessian
    a = SCHAModules.thermodynamic.w_to_a(w, self.current_T)
AttributeError: module 'SCHAModules' has no attribute 'thermodynamic'

I am attaching also the Ensemble.py

Davide

[Ensemble.py.txt](https://github.com/SSCHAcode/python-sscha/files/14022242/Ensemble.py.txt)

**Reply by diegomartinez2** [reply]
Looking at your Ensemble.py file I can see that the 'sed' command didn't changed the  "SCHAModules.thermodynamic" to "Thermodynamic".
Maybe you have to do it by hand? Edit the  "/gpfs/users/romanind/.conda/envs/Lorenzo_SSCHA/lib/python3.10/site-packages/sscha/Ensemble.py" to change that `a = SCHAModules.thermodynamic.w_to_a(w, self.current_T)` for `a = Thermodynamic.w_to_a(w, self.current_T)`.

**Reply by DromaninCM** [reply]
I changed it manually:

Traceback (most recent call last):
  File "/gpfs/workdir/romanind/CNT_SSCHA/hessian.py", line 48, in <module>
    dyn_hessian = ens.get_free_energy_hessian(include_v4 = INCLUDE_V4, get_full_hessian = True,verbose = False)
  File "/gpfs/users/romanind/.conda/envs/Lorenzo_SSCHA/lib/python3.10/site-packages/sscha/Ensemble.py", line 3657, in get_free_energy_hessian
    a = Thermodynamic.w_to_a(w, self.current_T)
NameError: name 'Thermodynamic' is not defined

I'm attaching the Ensemble.py

Davide

**Reply by Tongca** [reply]
I tried manually modifying the Ensemble.py file to change that SCHAModules.thermodynamic for Thermodynamic.  In addition, “import sscha.Thermodynamic as Thermodynamic” is added to the Ensemble.py header. The error disappears when compiled.
[Uploading Ensemble.txt…]()

**Comment by DromaninCM** [comment]
[Ensemble.py.txt](https://github.com/SSCHAcode/python-sscha/files/14022418/Ensemble.py.txt)

**Reply by Tongca** [reply]
Have you tried adding the line "import sscha.Thermodynamic as Thermodynamic" to the beginning of the Ensemble.py?

**Comment by DromaninCM** [comment]
> I tried manually modifying the Ensemble.py file to change that SCHAModules.thermodynamic for Thermodynamic. In addition, “import sscha.Thermodynamic as Thermodynamic” is added to the Ensemble.py header. The error disappears when compiled.
> Uploading Ensemble.txt…

Yes, I confirm the error now disappears also in my case.
I forgot to put the header to Thermodynamic in Ensemble manually

Davide

**Reply by mesonepigreco** [reply]
Very good! 
I selected Diego's answer and updated it to include all the missing steps. 
We will change this in the main repository and make these changes definitive, starting from the next minor release.
